### PR TITLE
Add task that can delete resources created by Istio

### PR DESCRIPTION
**Pull request by geoand**
This task is disabled by default.
This should be considered a crude first attempt at cleaning up the Istio
resources.

### CODE REVIEW

**Comment by geoand** [comment]
First fix for #21

File: ansible/istio/defaults/main.yml [hunk 1]
```diff
@@ -29,4 +25,7 @@ istio:
   bookinfo: false
 
   # Whether or not to open apps in the browser
-  open_apps: false
\ No newline at end of file
+  open_apps: false
+
+  # Whether to delete the old namespace and all data if it exists
+  delete_old: false
```
Comment (cmoulliard) [inline ansible/istio/defaults/main.yml]: Why do you name the param using delete_old ? Such param should be used to delete istio according to namespace param ....

Comment (geoand) [reply ansible/istio/defaults/main.yml]: This is just a flag that controls whether deleting will take place or not.
If this flag is set to true, then all the stuff under `istio.namespace` will be deleted.
If the flag is not set then no delete operations will take place before deploying Istio

Comment (cmoulliard) [reply ansible/istio/defaults/main.yml]: My question was more related to "_old". Why do you suffix the name with "_old" ?

**Comment by geoand** [comment]
For extra clarity basically.
A better name would probably be delete_existing.

Do you prefer we call it delete without any suffix?

**Comment by cmoulliard** [comment]
If we want to tell to the user what we will do then the name to be used should be `delete_resources`

**Comment by geoand** [comment]
I'll rename to `delete_resources` and update the PR

File: ansible/istio/tasks/delete_old.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  ignore_errors: true
+  with_items: "{{ istio_all_addons }}"
+
+- name: Delete Addon ServiceAccounts
+  command: "{{ oc_path }} delete sa istio-{{ item }}-service-account --ignore-not-found=true -n {{ istio.namespace }}"
+  ignore_errors: true
+  with_items: "{{ istio_all_addons }}"
+
+
+- name: Delete contect from Istio Kubernetes file
```
Comment (cmoulliard) [inline ansible/istio/tasks/delete_old.yml]: Typo error -> context

File: ansible/istio/tasks/delete_old.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  command: "{{ oc_path }} delete -f {{ istio_dir }}/install/kubernetes/istio.yaml --ignore-not-found=true -n {{ istio.namespace }}"
```
Comment (cmoulliard) [inline ansible/istio/tasks/delete_old.yml]: A problem that we will have with this command is that openshift will try to delete the namespace of istio-system (idem also for istio-auth)

Comment (geoand) [reply ansible/istio/tasks/delete_old.yml]: And just try to opportunistically remove things. What do you think?

Comment (cmoulliard) [reply ansible/istio/tasks/delete_old.yml]: Can we transform the file to remove the lines about namespace resource ?

Comment (geoand) [reply ansible/istio/tasks/delete_old.yml]: Yes, that's of course doable :)

**Comment by metacosm** [comment]
Just a quick comment: this commit actually mixes several concerns. It'd be better if commits were focused on doing just a single thing. This way it's easier to track (and revert, if needed) any issues.

**Comment by geoand** [comment]
Thank you for your comments Christophe!

Could you elaborate a little more please?

**Comment by metacosm** [comment]
For example, there are changes for the addons or install distro in this commit which don't seem related to the delete tasks. It would have been better, in my opinion, to separate these changes in different commits.

**Comment by geoand** [comment]
Thank you for the explanation!

I will go over it again and come back with my comments, because I can't
recall which commits went into which PR.

On Jan 22, 2018 10:28 PM, "Chris Laprun" <notifications@github.com> wrote:

For example, there are changes for the addons or install distro in this
commit which don't seem related to the delete tasks. It would have been
better, in my opinion, to separate these changes in different commits.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/snowdrop/istio-integration/pull/23#issuecomment-359554234>,
or mute the thread

**Comment by geoand** [comment]
I went through the code of the commit, and I remembered why I needed to change the add-on handling.

Basically when deleting the existing Istio installation we need to delete everything potentially keeping in mind that the configuration of existing installation is perhaps not the same as the one the user of playbook is using to perform the new Istio installations.
That it turn means that we need to attempt to delete all add-ons, why maintaining the ability to only install the add-ons that the user selected.

Furthermore the code get's complicated by the fact that we don't want to delete the whole namespace, but instead only the resources inside it (hence the copying and modification of the `istio.yaml` file). 

Although I am obviously biased, I believe that the commit really does only one thing, which is to provide the ability to delete existing resources. 
I am of course more than welcome to be persuaded that is not the case :)

**Pull request merged**